faunadb-java/src/main/java/com/faunadb/client/FaunaClient.java

```diff
@@ -330,23 +336,21 @@ private Value handleResponse(FullHttpResponse response) {
       return json.treeToValue(resource, Value.class);
     } catch (IOException ex) {
       throw new AssertionError(ex);
-    } finally {
-      response.release();
     }
   }
 
   private CompletableFuture<Value> performRequest(JsonNode body, Optional<Duration> queryTimeout) {
     try {
         return handleNetworkExceptions(connection.post("", body, queryTimeout).thenApply(this::handleResponse));
-    } catch (IOException ex) {
+    } catch (Exception ex) {
         CompletableFuture<Value> oops = new CompletableFuture<>();
         oops.completeExceptionally(ex);
         return oops;
     }
   }
 
-  private void handleQueryErrors(FullHttpResponse response) {
-    int status = response.status().code();
+  private void handleQueryErrors(HttpResponse<String> response) {
+    int status = response.statusCode();
     if (status >= 300) {
       try {
```

Comment:
The exception handling logic in this method is pretty problematic. It looks to me like this try block is overly broad: If parseResponseBody were rewritten to return a failure sentinel (could just be `null`), or throw a specific exception, rather than throwing a generic IOException, then the very unfortunate catch statement on what is now line 383 and also 385 could be removed.

Reply:
I am proposing the following patch https://github.com/fauna/faunadb-jvm/pull/270/commits/ad5f8358be7293721f2abccf67785a5ab402a0dd